keps/sig-node/20210706-pod-addmission-plugin.md

```diff
@@ -0,0 +1,314 @@
+capabilities of both the worker node and the container runtime.
+
+Kubelet will reject the Pod if any required capabilities in the Pod.Spec are not
+supported by the container engine running on the node. Such capabilities might
+include the ability to set sysctl parameters, use of elevated system privileges
+or use of a non-default process mount. Likewise, kubelet checks the Pod against
+node capabilities; for example, the presence of a specific apparmor profile or
+host kernel.
+
+These validations represent final, last-minute checks immediately before the Pod
+is started by the container runtime. These node-local checks differ from
+API-layer validations like Pod Security Policies or Validating Admission
+webhooks. Whereas the latter may be deactivated or removed by Kubernetes cluster
+administrators, the former node-local checks cannot be disabled.
+
+It is not currently possible to add additional validations before admitting the
+Pod. This document proposes a framework for enabling additional node-local Pod
+admission checks.
+
+## Motivation
+
+Amazon Elastic Kubernetes Service (EKS) provides users a managed Kubernetes
+control plane. EKS users are provisioned a Kubernetes cluster running on AWS
+cloud infrastructure. While the EKS user does not have host-level administrative
+access to the master nodes, it is important to point out that they do have
+administrative rights on that Kubernetes cluster.
+
+The EKS user’s worker node administrative access depends on the type of worker
+node the EKS user chooses. EKS users have three options. The first option is to
+bring their own EC2 instances as worker nodes. The second option is for EKS
+users to launch a managed worker node group. These first two options both result
+in the EKS user maintaining full host-level administrative rights on the worker
+nodes. The final option — the option that motivated this proposal — is for the
+EKS user to forego worker node management entirely using AWS Fargate, a
+serverless computing environment. With AWS Fargate, the EKS user does not have
+host-level administrative access to their worker node; in fact, the worker node
+runs on a serverless computing platform that abstracts away the entire notion of
+a host.
+
+In building the AWS EKS support for AWS Fargate, the AWS Kubernetes engineering
+team faced a dilemma: how could they prevent Pods destined to run on Fargate
+nodes from using host networking or assuming elevated host user privileges?
+
+The team initially investigated using a Pod Security Policy (PSP) that would
+prevent Pods with a Fargate scheduler type from having an elevated security
+context or using host networking. However, because the EKS user has
+administrative rights on the Kubernetes cluster, API-layer constructs such as a
+Pod Security Policy may be deleted, which would effectively disable the effect
+of that PSP. Likewise, the second solution the team landed on — using Node
```

Comment:
thank you for presenting in today's SIG Node meeting.
something that i found quite puzzling is the fact that you'd like the design to gate out on the cluster admin.
this was mentioned by `@Random-Liu` during the call.

not trusting the cluster admin is a peculiar case, since the cluster admin can already render a cluster inoperable, or change key RBAC roles.
can you elaborate more on why would you not truest them about potential disruptions, such as tempering with PSP removal or webhooks? why isn't documenting a best practice sufficient here?

Reply:
Thank you @neolit123 for giving quick read :) There are two risks as mentioned below that needs to be addressed,
1. Restrict pods which cannot execute on these managed worker nodes and 
2. Prevent pods from accessing some of the file system on the node or tampering other process on the host.

while `#1` can be solved with webhook or PSP and if cluster admins deletes them then it will impact their operations on this cluster, `#2` requires some validation in Kubelet before admitting the pods since cluster admins can delete webhook/PSP and schedule their pods onto these worker nodes directly by adding NodeName in their podSpec and perform elevated actions. Also, this bypasses the scheduler since cluster users populates the pod spec with nodename. I will elaborate in the KEP as well.

My thought here is, extending pod admit handler to support node-local plugins will help other providers who wanted to perform some last min checks in Kubelet.

Reply:
i can see the resulting complications if the cluster admins perform these actions, but it's still not clear to me why the platform would want to prevent them from doing that. is this about restricted scheduling such that can benefit the provider and not the cluster admin (in one form or the other)?

doing some high-level thinking here:
instead of adding more extension mechanisms, perhaps this should be considered as a problem in the priviliges ladder.
if there is a cluster-admin with full scope, there could be a persona on top of them that can lock certain objects.

this would overlap with the recent discussion about locking certain objects and blocking an accidental `kubectl delete <everything>`.

Reply:
> is this about restricted scheduling such that can benefit the provider and not the cluster admin (in one form or the other)?

Yes this is also one of our requirement that I called out earlier, `Prevent pods from accessing some of the file system on the node or tampering other process on the host`.

>  if there is a cluster-admin with full scope, there could be a persona on top of them that can lock certain objects.
Yes I like your thought process, something like blocking cluster-admin from deleting some objects for eg, webhook of certain types for example. This is where manifest based webhook would have really helped us but there were few cavets as Daniel pointed out - https://github.com/kubernetes/enhancements/pull/2074#issuecomment-703761276.